[Xtensa] Implement vararg support.

andreisfr opened the pull request.

andreisfr committed:
[Xtensa] Implement variable arguments support.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+    const TargetRegisterClass *RC = &Xtensa::ARRegClass;
+    MachineFrameInfo &MFI = MF.getFrameInfo();
+    MachineRegisterInfo &RegInfo = MF.getRegInfo();
+    unsigned RegSize = 4;
+    MVT RegTy = MVT::getIntegerVT(RegSize * 8);
```
Just use MVT::i32 directly?

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+    XtensaFI->setVarArgsFirstGPR(Idx + 2); // 2 - number of a2 register
+
+    XtensaFI->setVarArgsStackOffset(MFI.CreateFixedObject(
+        PtrVT.getSizeInBits() / 8, CCInfo.getStackSize(), true));
```
getStoreSize

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+      SDValue PtrOff = DAG.getFrameIndex(FI, getPointerTy(DAG.getDataLayout()));
+      SDValue Store = DAG.getStore(Chain, DL, ArgValue, PtrOff,
+                                   MachinePointerInfo::getFixedStack(MF, FI));
+      cast<StoreSDNode>(Store.getNode())
+          ->getMemOperand()
+          ->setValue((Value *)nullptr);
```
You shouldn't be adjusting this after the creation of the store

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+    // to the vararg save area.
+    for (unsigned I = Idx; I < ArgRegs.size(); ++I, VaArgOffset += RegSize) {
+      const unsigned Reg = RegInfo.createVirtualRegister(RC);
+      unsigned FrameReg = Subtarget.getRegisterInfo()->getFrameRegister(MF);
```
use Register, and pull out of loop

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -304,13 +318,14 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
   MachineFunction &MF = DAG.getMachineFunction();
   MachineFrameInfo &MFI = MF.getFrameInfo();
+  XtensaMachineFunctionInfo *XtensaFI = MF.getInfo<XtensaMachineFunctionInfo>();
+  EVT PtrVT = getPointerTy(MF.getDataLayout());
```
In context this probably should be getFrameIndexTy

andreisfr replied:
Fixed.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+      // Argument passed in FrameReg we save in A8 (in emitPrologue),
+      // so load argument from A8
+      if (ArgRegs[I] == FrameReg) {
+        RegInfo.addLiveIn(Xtensa::A8, Reg);
```
I don't see any reference to A8 in `emitPrologue`. It isn't live-in anyway.

andreisfr replied:
Thank you very much for comments. I removed redundant code (it will be used in later version).

andreisfr committed:
[Xtensa] Minor fixes.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -338,16 +329,13 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
       // Transform the arguments stored on
       // physical registers into virtual ones
-      unsigned Register = MF.addLiveIn(VA.getLocReg(), RC);
+      unsigned Register = MF.addLiveIn(VA.getLocReg(), &Xtensa::ARRegClass);
```
```suggestion
      Register Reg = MF.addLiveIn(VA.getLocReg(), &Xtensa::ARRegClass);
```
Use Register type and try to avoid shadowing type names with variable names

andreisfr replied:
Thank you very much for comments. Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -394,20 +382,18 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
-        Xtensa::A2, Xtensa::A3, Xtensa::A4, Xtensa::A5, Xtensa::A6, Xtensa::A7};
-    ArrayRef<MCPhysReg> ArgRegs = ArrayRef(XtensaArgRegs);
+    ArrayRef<MCPhysReg> ArgRegs = ArrayRef(IntRegs);
     unsigned Idx = CCInfo.getFirstUnallocated(ArgRegs);
```
Directly passing IntRegs to getFirstUnallocated should work I think

andreisfr replied:
Fixed

andreisfr committed:
[Xtensa] Minor fixes.

Fix variable names in XtensaMachineFunctionInfo class. Fix
LowerFormalArguments function.

arsenm commented on llvm/lib/Target/Xtensa/XtensaMachineFunctionInfo.h:
```diff
@@ -25,13 +25,14 @@ class XtensaMachineFunctionInfo : public MachineFunctionInfo {
-  int VarArgsStackOffset;
-  unsigned VarArgsFrameIndex;
+  unsigned VarArgsOnStackFrameIndex;
+  unsigned VarArgsInRegsFrameIndex;
```
```suggestion
  int VarArgsInRegsFrameIndex;
```

Frame indexes are signed

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaMachineFunctionInfo.h:
```diff
@@ -43,12 +44,18 @@ class XtensaMachineFunctionInfo : public MachineFunctionInfo {
 
-  int getVarArgsOnStackFrameIndex() const { return VarArgsStackOffset; }
-  void setVarArgsOnStackFrameIndex(int Offset) { VarArgsStackOffset = Offset; }
+  unsigned getVarArgsOnStackFrameIndex() const {
```
```suggestion
  int getVarArgsOnStackFrameIndex() const {
```

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaMachineFunctionInfo.h:
```diff
@@ -43,12 +44,18 @@ class XtensaMachineFunctionInfo : public MachineFunctionInfo {
+  unsigned getVarArgsOnStackFrameIndex() const {
+    return VarArgsOnStackFrameIndex;
+  }
+  void setVarArgsOnStackFrameIndex(unsigned FI) {
```
```suggestion
  void setVarArgsOnStackFrameIndex(int FI) {
```

andreisfr replied:
Thank you. Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaMachineFunctionInfo.h:
```diff
@@ -43,12 +44,18 @@ class XtensaMachineFunctionInfo : public MachineFunctionInfo {
   // Get and set the frame index of the first stack vararg.
-  unsigned getVarArgsInRegsFrameIndex() const { return VarArgsFrameIndex; }
-  void setVarArgsInRegsFrameIndex(unsigned FI) { VarArgsFrameIndex = FI; }
+  unsigned getVarArgsInRegsFrameIndex() const {
+    return VarArgsInRegsFrameIndex;
+  }
+  void setVarArgsInRegsFrameIndex(unsigned FI) { VarArgsInRegsFrameIndex = FI; }
```
```suggestion
  int getVarArgsInRegsFrameIndex() const {
    return VarArgsInRegsFrameIndex;
  }
  void setVarArgsInRegsFrameIndex(int FI) { VarArgsInRegsFrameIndex = FI; }
```

andreisfr replied:
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+
+    XtensaFI->setVarArgsFirstGPR(Idx + 2); // 2 - number of a2 register
+
+    XtensaFI->setVarArgsStackOffset(MFI.CreateFixedObject(
```
Fixed functions names.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+        Xtensa::A2, Xtensa::A3, Xtensa::A4, Xtensa::A5, Xtensa::A6, Xtensa::A7};
+    ArrayRef<MCPhysReg> ArgRegs = ArrayRef(XtensaArgRegs);
+    unsigned Idx = CCInfo.getFirstUnallocated(ArgRegs);
+    const TargetRegisterClass *RC = &Xtensa::ARRegClass;
```
I removed variable.

s-barannikov replied:
I can still see it, but whatever.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +393,68 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
 
+  if (IsVarArg) {
+    static const MCPhysReg XtensaArgRegs[6] = {
+        Xtensa::A2, Xtensa::A3, Xtensa::A4, Xtensa::A5, Xtensa::A6, Xtensa::A7};
```
Fixed.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -304,13 +318,14 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+  XtensaMachineFunctionInfo *XtensaFI = MF.getInfo<XtensaMachineFunctionInfo>();
+  EVT PtrVT = getPointerTy(MF.getDataLayout());
+
+  XtensaFI->setVarArgsFrameIndex(0);
```
Fxied.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -133,6 +134,14 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  // VASTART and VACOPY need to deal with the Xtensa-specific varargs
+  // structure, but VAEND is a no-op.
+  setOperationAction(ISD::VASTART, MVT::Other, Custom);
+  // we use special va_list structure so we have to customize this
```
Fixed.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -211,6 +220,11 @@ void XtensaTargetLowering::LowerAsmOperandForConstraint(
   TargetLowering::LowerAsmOperandForConstraint(Op, Constraint, Ops, DAG);
 }
 
+unsigned XtensaTargetLowering::getVaListSizeInBits(const DataLayout &DL) const {
```
Fixed.

andreisfr committed:
[Xtensa] Fix XtensaMachineFunctionInfo.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+
+  SDValue StkIndex =
+      DAG.getNode(ISD::ADD, DL, PtrVT, VAIndex,
+                  DAG.getConstant(32 + VASizeInBytes, DL, MVT::i32));
```
Maybe should use getObjectPtrOffset in a follow up for all of these adds

andreisfr replied:
I added use of getObjectPtrOffset to access vararg struct fields.

andreisfr committed:
[Xtensa] Minor fixes in lowering VASTART/VAARG.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  return DAG.getMemcpy(
+      Op.getOperand(0), Op, Op.getOperand(1), Op.getOperand(2),
+      DAG.getConstant(VAListSize, SDLoc(Op), MVT::i32), Align(4),
+      /*isVolatile=*/false, /*AlwaysInline=*/false,
```
```suggestion
      /*isVolatile=*/false, /*AlwaysInline=*/true,
```

andreisfr replied:
Fixed.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+      Op.getOperand(0), Op, Op.getOperand(1), Op.getOperand(2),
+      DAG.getConstant(VAListSize, SDLoc(Op), MVT::i32), Align(4),
+      /*isVolatile=*/false, /*AlwaysInline=*/false,
+      /*CI=*/nullptr, std::nullopt, MachinePointerInfo(), MachinePointerInfo());
```
VACOPY has provenance information, it should be propagated to memory operands.

andreisfr replied:
I corrected lowering VACOPY.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  const Value *SV = cast<SrcValueSDNode>(Node->getOperand(2))->getValue();
+  SDLoc DL(Node);
+  auto &TD = DAG.getDataLayout();
+  Align ArgAlignment = TD.getPrefTypeAlign(VT.getTypeForEVT(*DAG.getContext()));
```
Shouldn't this query ABI alignment instead?

andreisfr replied:
I changed to ABI alignment.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  unsigned ArgAlignInBytes = ArgAlignment.value();
+  unsigned ArgSizeInBytes =
+      TD.getTypeAllocSize(VT.getTypeForEVT(*DAG.getContext()));
+  unsigned VASizeInBytes = (ArgSizeInBytes + 3) & 0x3;
```
```suggestion
  unsigned VASizeInBytes = llvm::divideCeil(ArgSizeInBytes, 4);
```

andreisfr replied:
Thanks for comment, I corrected code.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,148 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  Align ArgAlignment = TD.getPrefTypeAlign(VT.getTypeForEVT(*DAG.getContext()));
+  unsigned ArgAlignInBytes = ArgAlignment.value();
+  unsigned ArgSizeInBytes =
+      TD.getTypeAllocSize(VT.getTypeForEVT(*DAG.getContext()));
```
Don't repeat getTypeForEVT twice. I also find it more confusing to use query the type alloc size and the alignment. If you need the explicit alignment, do the alignment yourself

andreisfr replied:
fixed

s-barannikov commented on llvm/test/CodeGen/Xtensa/vararg.ll:
```diff
@@ -0,0 +1,87 @@
+  call void @f(i32 %0)
+  br label %for.cond
+
+  call void @llvm.va_end(ptr %list)
```
It is unreachable.

andreisfr replied:
fixed

s-barannikov commented on llvm/test/CodeGen/Xtensa/vararg.ll:
```diff
@@ -0,0 +1,87 @@
+
+  call void @llvm.va_end(ptr %list)
+  ret void
+}
```
This needs more tests, e.g.
* some fixed arguments are passed in registers
* some fixed arguments are passed on the stack
* varargs with different types/sizes/alignments and their interaction with the above cases

andreisfr replied:
I added more tests.

andreisfr committed:
[Xtensa] Fix lowering VACOPY/VAARG.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -859,6 +909,155 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+    OrigIndex = DAG.getNode(ISD::ADD, DL, PtrVT, OrigIndex,
+                            DAG.getConstant(ArgAlignInBytes - 1, DL, MVT::i32));
+    OrigIndex = DAG.getNode(ISD::AND, DL, PtrVT, OrigIndex,
+                            DAG.getConstant(-ArgAlignInBytes, DL, MVT::i32));
```
This will trigger an assertion. Signed constants should be created with `getSignedConstant`.

andreisfr replied:
Thank you for comment. Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -378,6 +381,56 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
+        SDValue ArgValue = DAG.getCopyFromReg(Chain, DL, Reg, RegTy);
+        FI = MFI.CreateFixedObject(RegSize, VaArgOffset, true);
+        SDValue PtrOff =
+            DAG.getFrameIndex(FI, getPointerTy(DAG.getDataLayout()));
```
You should almost never use getPointerTy. This should be the frame index ty (and can be pulled out of the loop).

andreisfr replied:
Thank you for comment. Fixed.

andreisfr committed:
[Xtensa] Minor fixes.
